src/run/handlers/use-cache-handler.ts

```diff
@@ -0,0 +1,270 @@
+  // first released in v15.3.0-canary.13
+  CacheHandlerV2 as CacheHandler,
+} from 'next-with-cache-handler-v2/dist/server/lib/cache-handlers/types.js'
+
+import { getLogger } from './request-context.cjs'
+import {
+  getMostRecentTagRevalidationTimestamp,
+  isAnyTagStale,
+  markTagsAsStaleAndPurgeEdgeCache,
+} from './tags-handler.cjs'
+import { getTracer } from './tracer.cjs'
+
+// copied from default implementation
+// packages/next/src/server/lib/cache-handlers/default.ts
+type PrivateCacheEntry = {
+  entry: CacheEntry
+
+  // For the default cache we store errored cache
+  // entries and allow them to be used up to 3 times
+  // after that we want to dispose it and try for fresh
+
+  // If an entry is errored we return no entry
+  // three times so that we retry hitting origin (MISS)
+  // and then if it still fails to set after the third we
+  // return the errored content and use expiration of
+  // Math.min(30, entry.expiration)
+
+  isErrored: boolean // pieh: this doesn't seem to be actually used in the default implementation
+  errorRetryCount: number // pieh: this doesn't seem to be actually used in the default implementation
+
+  // compute size on set since we need to read size
+  // of the ReadableStream for LRU evicting
+  size: number
+}
+
+type CacheHandleLRUCache = LRUCache<string, PrivateCacheEntry>
+type PendingSets = Map<string, Promise<void>>
+
+const LRU_CACHE_GLOBAL_KEY = Symbol.for('nf-use-cache-handler-lru-cache')
+const PENDING_SETS_GLOBAL_KEY = Symbol.for('nf-use-cache-handler-pending-sets')
+const cacheHandlersSymbol = Symbol.for('@next/cache-handlers')
+const extendedGlobalThis = globalThis as typeof globalThis & {
+  // Used by Next Runtime to ensure we have single instance of
+  //  - LRUCache
+  //  - pending sets
+  // even if this module gets copied multiple times
+  [LRU_CACHE_GLOBAL_KEY]?: CacheHandleLRUCache
+  [PENDING_SETS_GLOBAL_KEY]?: PendingSets
+
+  // Used by Next.js to provide implementation of cache handlers
+  [cacheHandlersSymbol]?: {
+    RemoteCache?: CacheHandler
+    DefaultCache?: CacheHandler
+  }
+}
+
+function getLRUCache(): CacheHandleLRUCache {
+  if (extendedGlobalThis[LRU_CACHE_GLOBAL_KEY]) {
+    return extendedGlobalThis[LRU_CACHE_GLOBAL_KEY]
+  }
+
+  const lruCache = new LRUCache<string, PrivateCacheEntry>({
+    max: 1000,
+    maxSize: 50 * 1024 * 1024, // same as hardcoded default in Next.js
+    sizeCalculation: (value) => value.size,
+  })
+
+  extendedGlobalThis[LRU_CACHE_GLOBAL_KEY] = lruCache
+
+  return lruCache
+}
+
+function getPendingSets(): PendingSets {
+  if (extendedGlobalThis[PENDING_SETS_GLOBAL_KEY]) {
+    return extendedGlobalThis[PENDING_SETS_GLOBAL_KEY]
+  }
+
+  const pendingSets = new Map()
+  extendedGlobalThis[PENDING_SETS_GLOBAL_KEY] = pendingSets
+  return pendingSets
+}
+
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+const tmpResolvePendingBeforeCreatingAPromise = () => {}
```

Comment:
This is our temp variable, or was this also copy pasted?

Reply:
it was originally same as in  https://github.com/vercel/next.js/blob/0eb78fd629858858a5a31591b3b3de294762f870/packages/next/src/server/lib/cache-handlers/default.ts#L101, just our lintings rule didn't like defining this temporary no-op function in the same place:

![image](https://github.com/user-attachments/assets/4cf3cbfd-3562-4533-a8fe-5de1604f2a95)

and to me, the rule mede sense so I did move it to outer scope

the purpose of tmp function there that we re-assign in the `new Promise` callback in the first place is to make typescript happy, because otherwise `resolvePending` is either nullable or we have to force a type, so none of the options are really great and I just thought the way it currently is was the least worst option